### DISCUSSION TITLE
Act warning using Link white testing with Remix Testing

### DISCUSSION THREAD

**Discussion by Afsoon** [remix-run/remix#7555] [Q&A] [answered]
Hello everyone,

This warning is probably one reason that `@remix-run/testing` is still unstable. We are using `@remix-run/testing` for unit testing of our pages or components and we notice that when you change between routes, `vitest` is emitting the following warning:
```
Warning: An update to RouterProvider inside a test was not wrapped in act(...).

When testing, code that causes React state updates should be wrapped into act(...):

act(() => {
  /* fire events that update state */
});
/* assert on the output */

This ensures that you're testing the behavior the user would see in the browser. Learn more at https://reactjs.org/link/wrap-tests-with-act
    at RouterProvider (/Users/said/Documents/poc/node_modules/react-router/dist/umd/react-router.development.js:898:7)
    at RemixStub (/Users/said/Documents/poc/node_modules/@remix-run/testing/dist/create-remix-stub.js:42:5)
```

I ran out of ideas to overcome the error without using `act`. I'd like to know if we must use `act` until the library is 100% stable or if there is a way to suppress the error. A minimal reproduction to have this warning is the following snippet of code:

```typescript
it("should render overview page", async () => {
    const user = userEvent.setup();
    const RemixStub = createRemixStub([
      {
        path: "/",
        id: "root",
        Component: () => <Link to={"/overview"}>Go to Overview</Link>,
      },
      {
        path: "/overview",
        id: "overview",
        Component: () => <div>Overview</div>,
      },
    ]);
    render(<RemixStub initialEntries={["/"]} />);

    await user.click(screen.getByRole("link", { name: "Go to Overview" }));
    expect(screen.queryByText("Go to Overview")).not.toBeInTheDocument();
    expect(screen.getByText("Overview")).toBeInTheDocument();
  })
```

**Comment by Afsoon** [comment]
I don't know if I'm correct, but I found the reason we should wrap in an `act` the `user.click()`. Under the hood, we are fetching the loader from the route `/overview` asynchronous, and it's happening outside my test code.

If you try this code using `happy-dom` pre `12.2.1`, it will fail, but version `12.2.1` won't fail. Why am I so specific with the version? My source is because I fixed this bug in `happy-dom` a few days ago, https://github.com/capricorn86/happy-dom/issues/1097

**Reply by Afsoon** [reply]
@brophdawg11 @MichaelDeBoey any though on this?

**Reply by brophdawg11** [reply, marked as answer]
This warning doesn't come from the remix testing library, rather from underlying react testing code.  Act is opt-in as of react 18 so I would check your vitest configuration and your react testing versions and see if one of those is opting into the warning.

https://react.dev/blog/2022/03/08/react-18-upgrade-guide

This link provides some good information on the purpose of act and when you should need it:

https://github.com/reactwg/react-18/discussions/102

**Reply by Afsoon** [reply]
Sorry I forgot to answer. Thank you for the information @brophdawg11, I think theses changes made by React Team get unnoticed by my side.
